### ISSUE TITLE
Intel Compiler:  CMake C/Fortran interface linking

### ISSUE THREAD

**Issue by jrood-nrel** [trilinos/Trilinos#7740]
## Question

I am using the Intel Compiler 2019.5. While trying to build Trilinos, I notice CMake will check mixed language C/Fortran linking is working before allowing the configure to finish. However, it always tries to link with the C++ compiler instead of the Fortran compiler, causing the error below. I have tried changing the precedence in CMake using https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_LINKER_PREFERENCE.html#variable:CMAKE_<LANG>_LINKER_PREFERENCE but switching the default values from 20 <-> 30 for C++ and Fortran to give the Fortran linker precedence does not seem to work.

In short, the question is really, does anybody have a workaround for using Intel to build Trilinos? This seems to be mostly CMake's fault in my opinion though.

```
compilers_and_libraries_2019.5.281/linux/mpi/intel64/bin/mpiicpc  -O3 -DNDEBUG   CMakeFiles/VerifyFortranC.dir/main.c.o CMakeFiles/VerifyFortranC.dir/VerifyC.c.o CMakeFiles/VerifyFortranC.dir/VerifyCXX.cxx.o  -o VerifyFortranC  libVerifyFortran.a
     499        libVerifyFortran.a(VerifyFortran.f.o): In function `verifyfortran_':
  >> 500    VerifyFortran.f:(.text+0x3b): undefined reference to `for_write_seq_lis'
```

**Comment by jhux2** [comment]
@bartlettroscoe @trilinos/framework

**Comment by jrood-nrel** [comment]
As with all things Intel MPI 2019 related. This is actually a libfabric issue. It turns out CMake can't figure out the correct thing to do for this linking because its introspection fails when it can't find libfabric in `LD_LIBRARY_PATH`. Someone having this identical problem is https://github.com/spack/spack/issues/15522 . Since I am using Spack, I manually included the `LD_LIBRARY_PATH` (since Spack will filter it out unless it gets populated inside Spack itself) to libfabric included by Intel MPI in the Trilinos `package.py` as below. Currently I didn't have a better general option than hardcoding this path to solve this specific issue.
```
    def setup_build_environment(self, env):
        if '%intel' in self.spec:
            env.append_path('LD_LIBRARY_PATH', join_path(self.spec['mpi'].prefix, 'compilers_and_libraries_2020.1.217', 'linux', 'mpi', 'intel64', 'libfabric', 'lib'))
            env.append_path('FI_PROVIDER_PATH', join_path(self.spec['mpi'].prefix, 'compilers_and_libraries_2020.1.217', 'linux', 'mpi', 'intel64', 'libfabric', 'lib', 'prov'))
```

**Comment by hpcnpatel** [comment]
FYI, I still face the error with spack 15.3 
```
  >> 50    CMake Error at /dss/dsshome1/09/di36pex/SPACK_lrz_dir/master-base-0150/opt/haswell/cmake/3.16.5-gcc-76nzofy/share/cmake-3.16/Modules/CMake
           TestFortranCompiler.cmake:45 (message):
     51      The Fortran compiler
     52    
     53        "/dss/dsshome1/09/di36pex/SPACK_lrz_dir/master-base-0150/opt/haswell/intel-mpi/2019.7.217-gcc-3m4egml/compilers_and_libraries_2020.1.2
           17/linux/mpi/intel64/bin/mpif90"
     54    
     55      is not able to compile a simple test program.
```
Not only for Trilinos, but I have also observed this signature with some other packages. Seems that complete MPI & MKL paths are not propagated through CMake.

**Comment by hpcnpatel** [comment]
similar work around is,
```
spack_env.prepend_path('LD_LIBRARY_PATH',self.component_lib_dir('mpi')+'/../libfabric/lib')

```
